Components: Add experimental `ConfirmDialog` (#34153)
WordPress/gutenberg: packages/components/src/confirm-dialog/README.md

# `ConfirmDialog`

<div class="callout callout-alert">
This feature is still experimental. "Experimental" means this is an early implementation subject to drastic and breaking changes.
</div>

`ConfirmDialog` displays a confirmation dialog as a `Modal`, with `OK` and `Cancel` buttons. It's confirmed by clicking `OK` or by pressing the `Enter` key. It's cancelled by clicking `Cancel` or by pressing the `ESC` key or clicking outside the dialog focus (i.e, the overlay).

> It's built on top of `Modal`, so you can override any of the `Modal` props if you'd like. This also means that it's added as a child of the `body` element in the DOM tree through a React `Portal`, regardless of where it's declared in the React tree.

Reviewer: A few changes:
- Let's add a link to the `Modal` component
- let's remove the part about React Portal — this component's docs should not mention another component's implementation details
- Let's not quote the exact text on the confirm and cancel buttons, as the text may change in the future

```suggestion
`ConfirmDialog` is built of top of [`Modal`](/packages/components/src/modal/README.md] and displays a confirmation dialog, with _confirm_ and _cancel_ buttons.

The dialog is confirmed by clicking the _confirm_ button or by pressing the `Enter` key. It is cancelled (closed) by clicking the _cancel_ button, by pressing the `ESC` key, or by clicking outside the dialog focus (i.e, the overlay).
```

Author: Yeah, good point on the implementation details too! Thanks! Now that I think of it, I was also kinda rubber-🦆'ing with myself as I learned more about the components and tried to figure out the implementation.

## Usage

`ConfirmDialog` has two main implicit modes: controlled and uncontrolled.

### Uncontrolled mode

Allows the component to be used standalone, just by declaring it as part of another React's component render method:
  * It will be automatically open (displayed) upon mounting;

>   * It will automatically close itself when the `x`, `Cancel`, `Confirm` or overlay is clicked;

Reviewer: Similarly to what discussed before:

```suggestion
  * It will be automatically closed when when clicking the _cancel_ button, by pressing the `ESC` key, or by clicking outside the dialog focus (i.e, the overlay);
```

  * `onCancel` is not mandatory but can be passed. Even if passed, the dialog will still be able to close itself.

Activating this mode is as simple as omitting the `isOpen` prop. The only mandatory prop, in this case, is the `onConfirm` callback.

```jsx
import {
  __experimentalConfirmDialog as ConfirmDialog
} from '@wordpress/components';

function Example() {
  return (
    <ConfirmDialog onConfirm={ () => console.debug(' Confirmed! ') }>

>   )

Reviewer 2: ```suggestion
  return (
    <ConfirmDialog onConfirm={ () => console.debug(' Confirmed! ') } />
  );
```

Also, if `message` is required, we should add it here.

}
```

### Controlled mode

Let the parent component control when the dialog is open/closed. It's activated when a boolean value is passed to `isOpen`:
  * It will not be automatically closed. You need to let it know when to open/close by updating the value of the `isOpen` prop;
  * Both `onConfirm` and the `onCancel` callbacks are mandatory props in this mode;
  * You'll want to update the state that controls `isOpen` by updating it from the `onCancel` and `onConfirm` callbacks.

```jsx
import {
  __experimentalConfirmDialog as ConfirmDialog
} from '@wordpress/components';

function Example() {
  const [ isOpen, setIsOpen ] = useState( true );

  const handleConfirm = () => {
    console.debug( 'Confirmed!' );
    setIsOpen( false );
  }

  const handleCancel = () => {
    console.debug( 'Cancelled!' );
    setIsOpen( false );
  }

  return (
    <ConfirmDialog isOpen={ isOpen } onConfirm={ handleConfirm } onCancel={ handleCancel }>

>   )

Reviewer 2: ```suggestion
  return (
    <ConfirmDialog
      isOpen={ isOpen }
      onConfirm={ handleConfirm }
      onCancel={ handleCancel }
    />
  );
```
Also, if `message` is required, we should add it here.

Author: Ah yeah, forgot to update it. Thanks for pointing that out!

}
```

### Unsupported: Multiple instances

> Multiple `ConfirmDialog's is an edge case that's currently not officially supported by this component. At the moment, new instances will end up closing the last instance due to the way the `Modal` is implemented (more specifically, because of the use of the `onFocusOutside` to detect a blur in order to close it). This will end up automatically cancelling the previous instance, which might or might not be what you want.

Reviewer: On the same note as what was said above — it's probably a good idea not to be too specific about another component's implementation details.

```suggestion
Multiple `ConfirmDialog's is an edge case that's currently not officially supported by this component. At the moment, new instances will end up closing the last instance due to the way the `Modal` is implemented.
```

> Suppose you need to handle multiple confirmations without discarding other instances. In that case, it might be better to have a singleton wrapper component that provides a context-based API to trigger the dialog, keeping track of multiple instances. Here's an [example](https://github.com/WordPress/gutenberg/pull/34153#issuecomment-908342367).

Reviewer: Should we add a Storybook example with this scenario?

Author: I think we should probably remove this for now as this won't be included as part of this changeset.

Author: Done, removed this paragraph for now.

## Custom Types

```ts
type DialogInputEvent =
	| KeyboardEvent< HTMLDivElement >
	| MouseEvent< HTMLButtonElement >
```

> ## Props

Reviewer 2: `message` is missing.

function Example() {
  const [ isOpen, setIsOpen ] = useState( true );

  const handleConfirm = () => {
    console.debug( 'Confirmed!' );
    setIsOpen( false );
  }

  const handleCancel = () => {
    console.debug( 'Cancelled!' );
    setIsOpen( false );
  }

  return (
    <ConfirmDialog isOpen={ isOpen } onConfirm={ handleConfirm } onCancel={ handleCancel }>
  )
}
```

### Unsupported: Multiple instances

Multiple `ConfirmDialog's is an edge case that's currently not officially supported by this component. At the moment, new instances will end up closing the last instance due to the way the `Modal` is implemented (more specifically, because of the use of the `onFocusOutside` to detect a blur in order to close it). This will end up automatically cancelling the previous instance, which might or might not be what you want.

Suppose you need to handle multiple confirmations without discarding other instances. In that case, it might be better to have a singleton wrapper component that provides a context-based API to trigger the dialog, keeping track of multiple instances. Here's an [example](https://github.com/WordPress/gutenberg/pull/34153#issuecomment-908342367).

> ### Props

Reviewer: ```suggestion
## Props
```

### `isOpen`: `boolean`

Defines if the dialog is open (displayed) or closed (not rendered/displayed). It also implicitly toggles the controlled mode if set or the uncontrolled mode if it's not set.

### `onConfirm`: `( event: SyntheticEvent ) => void`

- Required: Yes

The callback that's called when the user confirms. A confirmation can happen when the `OK` button is clicked or when `Enter` is pressed.

### `onCancel`: `(event: SyntheticEvent ) => void`

> - Required: Yes if `isOpen` is set, No if `isOpen` is not set

Reviewer: > - Required: Yes if `isOpen` is set, No if `isOpen` is not set

Is there any way we can improve this situation, so that we don't have conditionally required props? Maybe the best solution is to make it always required?

Or to expose two different components, one controlled and one uncontrolled? @diegohaz what do you think?

Reviewer: We could also try experimenting with TypeScript types, something like

```ts
type typeA = {
  isOpen?: boolean;
  onCancel?: (event: SyntheticEvent ) => void;
};

// `onCancel` is required if `isOpen` is set
type typeB = {
  isOpen: boolean;
  onCancel: (event: SyntheticEvent ) => void;
};

type union = typeA | typeB;
```

Not sure.

Author: Yeah, this is the part I want the most feedback on. I came up with this API as I built the component to look and behave like the native `confirm` component (and be easy and straightforward to use), while still allowing it to be used as a controlled component. I think it mainly depends on how we plan on consuming this component:

* If we think it's fine to just use it uncontrolled inline in the components that need it, then yes, we should keep this or break into two components as you said;
* if we'll instead add more infrastructure around it (like what Diego suggested [here](https://github.com/WordPress/gutenberg/pull/34153#issuecomment-908342367)) maybe it doesn't make sense to keep the uncontrolled aspect.

Reviewer: Hey @fullofcaffeine , sorry I haven't had much time to follow up on this. Have you made any changes to how the controlled vs uncontrolled components are exposed, or to how the variables are typed?

Reviewer: This [comment](https://github.com/WordPress/gutenberg/pull/34153#discussion_r727166119) may actually include a good solution for typing this!
